communication/mavlink_waypoint_handler.cpp

```diff
@@ -517,1514 +385,285 @@ void Mavlink_waypoint_handler::clear_waypoint_list(Mavlink_waypoint_handler* way
-               if (navigation_.dist2wp_sqr <= waypoint_radius)
-                {
-                    result = true;
-                }
-            break;
-
-            case Navigation::strategy_t::DUBIN:
-                if (state_.autopilot_type == MAV_TYPE_QUADROTOR)
-                {
-                    if (navigation_.dist2wp_sqr <= waypoint_radius)
-                    {
-                        result = true;
-                    }
-                }
-                else
-                {
-                    if (ins_.position_lf()[Z] <= navigation_.takeoff_altitude)
-                    {
-                        result = true;
-                    }
-                }
-            break;
-        }
-
-        if (result)
-        {
-            navigation_.dubin_state = DUBIN_INIT;
-
-            if (!state_.nav_plan_active)
-            {
-                waypoint_coordinates_ = waypoint_hold_coordinates;
-                waypoint_coordinates_.radius = 0.0f;
-            }
-
-            print_util_dbg_print("Automatic take-off finished.\r\n");
-        }
+        return waypoint_list_[current_waypoint_index_];
+    }
+    else // TODO: Return an error
+    {
+        // For now, return last waypoint structure
+        return waypoint_list_[waypoint_count_-1];
     }
-
-    return result;
 }
 
-mav_result_t Mavlink_waypoint_handler::start_stop_navigation(Mavlink_waypoint_handler* waypoint_handler, mavlink_command_long_t* packet)
+const Waypoint& Mavlink_waypoint_handler::next_waypoint()
 {
-    mav_result_t result = MAV_RESULT_UNSUPPORTED;
+    // If there are no waypoints set, create hold position
+    if (waypoint_count_ == 0)
+    {
+        waypoint_list_[0] = Waypoint(   MAV_FRAME_LOCAL_NED,
+                                        MAV_CMD_NAV_LOITER_UNLIM,
+                                        0,
+                                        0.0f,
+                                        0.0f,
+                                        0.0f,
+                                        0.0f,
+                                        ins_.position_lf()[X],
+                                        ins_.position_lf()[Y],
+                                        ins_.position_lf()[Z]);
+        waypoint_count_ = 1;
+        current_waypoint_index_ = 0;
+        return waypoint_list_[0];
+    }
 
-    if (packet->param1 == MAV_GOTO_DO_HOLD)
+    // If it is a good index
+    if (current_waypoint_index_ >= 0 && current_waypoint_index_ < waypoint_count_)
     {
-        if (packet->param2 == MAV_GOTO_HOLD_AT_CURRENT_POSITION)
+        // Check if the next waypoint exists
+        if ((current_waypoint_index_ + 1) != waypoint_count_)
         {
-            waypoint_handler->hold_init(waypoint_handler->ins_.position_lf());
-
-            waypoint_handler->navigation_.internal_state_ = Navigation::NAV_STOP_ON_POSITION;
-
-            result = MAV_RESULT_ACCEPTED;
+            return waypoint_list_[current_waypoint_index_ + 1];
```

Comment:
If the current waypoint is not autocontinue, the handlers will not allow the mission planner to advance to the next waypoint. Something needs to be decided about what the drone should do next at the end of the mission: go back and start again, go home, stay at the last mission waypoint, or something else. Currently, the code is written to go back to the beginning.